### ISSUE TITLE
Why cannot the ransac be integrated with the correspondence prediction network?

### ISSUE THREAD

**Issue by xztao** [cvlab-dresden/DSAC#19]
I am reading the DSAC paper these days and got confused for the problem.
The motivation of DSAC is to find a workaround to improve the undifferentiable Ransac, for the e2e training of pose regression network. The undifferentiable step in Ransac pipeline is the argmax, as illustrated in the paper. But can it be simply treated as what max pooling layer does in back propagation step?
![image](https://github.com/cvlab-dresden/DSAC/assets/99629054/cb8c151f-f671-4588-81e5-c668cbd22f05)

As an example, We sample 3 correspondence subset, named Y_a, Y_b and Y_c. Then we calculate the score via PnP, say S_a, S_b and S_c. Assume the maximum is S_a. Then that find the maximum is not differentiable indeed, like max pooling. But the derivative of the loss w.r.t. Y_a can be easily obtained to be exploited in the derivative chain.

I wonder where is wrong in my thought.. Thanks a lot for any help!

**Comment by ebrach** [comment]
Hi!

I think what you describe is a possible alternative, but I'm not sure how well it would work. The learning signal will be much sparser than the signal provided by DSAC. In DSAC, we sample 256 pose hypotheses and all of them contribute to the final loss. Each hypothesis' score is also considered in learning: Bad hypotheses should have low scores and good hypotheses should have high scores - such that we are able to reliably find the best hypotheses at test time.

In the solution you propose, from all 256 hypotheses only one is providing gradients for learning, which might be unstable. In your formulation there is also no signal to inform learning the hypothesis score, despite you needing the score to determine which hypothesis is the best one.

Best,
Eric

**Comment by ebrach** [comment]
Closing this as an issue, since it's not code related. But feel free to continue commenting.

**Comment by xztao** [comment]
> Hi!
> 
> I think what you describe is a possible alternative, but I'm not sure how well it would work. The learning signal will be much sparser than the signal provided by DSAC. In DSAC, we sample 256 pose hypotheses and all of them contribute to the final loss. Each hypothesis' score is also considered in learning: Bad hypotheses should have low scores and good hypotheses should have high scores - such that we are able to reliably find the best hypotheses at test time.
> 
> In the solution you propose, from all 256 hypotheses only one is providing gradients for learning, which might be unstable. In your formulation there is also no signal to inform learning the hypothesis score, despite you needing the score to determine which hypothesis is the best one.
> 
> Best, Eric

Happy to receive your reply! I recognized from your reply that the DSAC is more like someway to transform the 256 hypotheses into gradient descent signal comprehensively. If it makes sense, I think the network of scoring part can be replaced with the standard scoring function of Ransac, such as the inlier ratio( inlier_num/correspondence_num ). And then make a softmax distribution for the 256 scores. It can also tell whether the correspondence prediction weights are good or not, as a learning signal.

**Comment by ebrach** [comment]
Indeed, but it has to be a soft inlier count to support back-prob. That is what we did in a follow-up paper for DSAC++:
https://openaccess.thecvf.com/content_cvpr_2018/papers/Brachmann_Learning_Less_Is_CVPR_2018_paper.pdf

**Comment by xztao** [comment]
> Indeed, but it has to be a soft inlier count to support back-prob. That is what we did in a follow-up paper for DSAC++: https://openaccess.thecvf.com/content_cvpr_2018/papers/Brachmann_Learning_Less_Is_CVPR_2018_paper.pdf

Cool job! I will read it later. Have a nice day, Eric!
